backends/dpdk/dpdkArch.cpp

```diff
@@ -2877,4 +2879,174 @@ const IR::Node* ElimHeaderCopy::postorder(IR::Member* m) {
+        if (structure->header_type == st->name.name)
+            return true;
+        else {
+            for (auto f : st->fields) {
+                cstring fname = f->name.name;
+                // field name "f0" is for detecting automatically generated struct, used for
+                // initializing with structexpression
+                if (f->type->is<IR::Type_Header>() && fname != "f0")
+                    return true;
+                else if (auto tn = f->type->to<IR::Type_Name>()) {
+                    auto type0 = typeMap->getTypeType(tn, true);
+                    return type0->is<IR::Type_Header>() && fname != "f0";
+                } else {
+                    return false;
+                }
+            }
+        }
+    }
+    return true;
+}
+
+const IR::Node* DpdkAddPseudoHeaderDecl::preorder(IR::P4Control* c) {
+    if (is_all_args_header) return c;
+    auto ctrl = getOriginal<IR::P4Control>();
+    for (auto p : ctrl->type->applyParams->parameters) {
+        if (auto tn = p->type->to<IR::Type_Name>()) {
+            auto decl = typeMap->getTypeType(tn, true);
+            if (isHeadersStruct(decl->to<IR::Type_Struct>())) {
+                headersDeclName = p->name.name;
+            }
+        }
+    }
+    return c;
+}
+
+std::pair<IR::AssignmentStatement*, IR::Member*>
+MoveNonHeaderFieldsToPseudoHeader::addAssignmentStmt(const IR::NamedExpression* ne) {
+    auto name = refMap->newName("pseudo");
+    BUG_CHECK(typeMap->getType(ne->expression, true)->is<IR::Type_Bits>(), "Unexpected type");
+    auto type = typeMap->getType(ne->expression, true)->to<IR::Type_Bits>();
+    auto width = (type->width_bits() + 7) & (~7);
+    pseudoFieldNameType.push_back(
+        std::pair<cstring, const IR::Type*>(name, IR::Type_Bits::get(width)));
+    auto mem0 =
+        new IR::Member(new IR::PathExpression(IR::ID(DpdkAddPseudoHeaderDecl::headersDeclName)),
+                       IR::ID(headerInstanceName));
+    auto mem1 = new IR::Member(mem0, IR::ID(name));
+    auto cast1 = new IR::Cast(IR::Type_Bits::get(width), ne->expression);
+    return {new IR::AssignmentStatement(mem1, cast1), mem1};
+}
+
+const IR::Node* MoveNonHeaderFieldsToPseudoHeader::postorder(IR::MethodCallStatement* statement) {
+    if (is_all_args_header) return statement;
+    auto mce = statement->methodCall;
+    bool added_copy = false;
+    IR::Type_Name* newTname = nullptr;
+    IR::StructExpression* struct_exp = nullptr;
+    auto result = new IR::IndexedVector<IR::StatOrDecl>();
+    if (auto* m = mce->method->to<IR::Member>()) {
+        if (auto* type = typeMap->getType(m->expr)->to<IR::Type_Extern>()) {
+            if (type->name == "InternetChecksum") {
+                if (m->member == "add" || m->member == "subtract") {
+                    auto components = new IR::IndexedVector<IR::NamedExpression>();
+                    for (auto arg : *mce->arguments) {
+                        BUG_CHECK(arg->expression->is<IR::StructExpression>(),
```

Comment:
I think there is no restriction on the arguments to be StructExpression. Constants, PathExpressions, Member expressions should be allowed as well.

Reply:
done